File: ios-SpaceSub/ios-SpaceSub/Models/Analytics.swift
import Foundation

// MARK: - Legacy (backward compat)

nonisolated struct AnalyticsResponse: Codable, Sendable {
    let monthlyTotal: Double
    let yearlyTotal: Double
    let activeSubscriptions: Int
    let topSubscriptions: [TopSubscription]
    let upcomingCharges: [UpcomingCharge]
}

nonisolated struct TopSubscription: Codable, Identifiable, Sendable {
    let merchant: String
    let amount: Double
    let periodType: BillingCycle
    let monthlyEquivalent: Double

    var id: String { merchant }
}

nonisolated struct UpcomingCharge: Codable, Identifiable, Sendable {
    let merchant: String
    let amount: Double
    let nextExpectedCharge: String
    let periodType: BillingCycle

    var id: String { "\(merchant)-\(nextExpectedCharge)" }
}

// MARK: - Overview

nonisolated struct AnalyticsOverview: Codable, Sendable {
    let mrr: Double
    let arr: Double
    let activeCount: Int
    let upcomingCount: Int
    let periodTotal: Double
    let trend: TrendInfo
}

nonisolated struct TrendInfo: Codable, Sendable {
    let currentMonth: Double
    let prevMonth: Double
    let changePct: Double
}

// MARK: - By Category

nonisolated struct CategoryItem: Codable, Identifiable, Sendable {
    let category: String
    let color: String
    let total: Double
    let count: Int
    let percent: Double

    var id: String { category }
}

// MARK: - By Service

nonisolated struct ServiceItem: Codable, Identifiable, Sendable {
    let merchant: String
    let monthlyAmount: Double
    let yearlyAmount: Double
    let category: String
    let color: String

    var id: String { merchant }
}

// MARK: - By Period

nonisolated struct PeriodItem: Codable, Identifiable, Sendable {
    let period: String
    let total: Double
    let count: Int
    let momGrowthPct: Double?

    var id: String { period }
}

// MARK: - Scores

nonisolated enum ChurnRisk: String, Codable, Sendable {
    case LOW, MEDIUM, HIGH
}

nonisolated struct ScoreItem: Codable, Identifiable, Sendable {
    let subscriptionId: String
    let merchant: String
    let valueScore: Int
    let churnRisk: ChurnRisk
    let label: String
    let monthlyAmount: Double

    var id: String { subscriptionId }
}

// MARK: - Recommendations

nonisolated enum RecoType: String, Codable, Sendable {
    case CANCEL, REVIEW, DOWNGRADE, CONSOLIDATE
}

nonisolated enum RecoPriority: String, Codable, Sendable {
    case HIGH, MEDIUM, LOW
}

nonisolated struct RecommendationItem: Codable, Identifiable, Sendable {
    let type: RecoType
    let priority: RecoPriority
    let merchant: String
    let currentCost: Double
    let potentialSavings: Double
    let reason: String

    var id: String { "\(type.rawValue)-\(merchant)" }
}

// MARK: - Period with Moving Average

struct PeriodItemWithAvg: Identifiable {
    let id: String
    let period: String
    let total: Double
    let count: Int
    let momGrowthPct: Double?
    let movingAvg: Double
}

// MARK: - Score Filter

enum ScoreFilter: String, CaseIterable, Identifiable {
    case all, risky, healthy

    var id: String { rawValue }

    var label: String {
        switch self {
        case .all: "Все"
        case .risky: "Проблемные"
        case .healthy: "Здоровые"
        }
    }
}

File: android-SpaceSub/app/src/main/java/dev/squad52/spacesub/models/Analytics.kt
package dev.squad52.spacesub.models

import com.google.gson.annotations.SerializedName

data class AnalyticsOverview(
    val mrr: Double,
    val arr: Double,
    val activeCount: Int,
    val upcomingCount: Int,
    val periodTotal: Double,
    val trend: TrendInfo
)

data class TrendInfo(
    val currentMonth: Double,
    val prevMonth: Double,
    val changePct: Double
)

data class CategoryItem(
    val category: String,
    val color: String,
    val total: Double,
    val count: Int,
    val percent: Double
)

data class ServiceItem(
    val merchant: String,
    val monthlyAmount: Double,
    val yearlyAmount: Double,
    val category: String,
    val color: String
)

data class PeriodItem(
    val period: String,
    val total: Double,
    val count: Int,
    val momGrowthPct: Double? = null
)

enum class ChurnRisk {
    @SerializedName("LOW") LOW,
    @SerializedName("MEDIUM") MEDIUM,
    @SerializedName("HIGH") HIGH
}

data class ScoreItem(
    val subscriptionId: String,
    val merchant: String,
    val valueScore: Double,
    val churnRisk: ChurnRisk,
    val label: String,
    val monthlyAmount: Double
)

enum class RecoType {
    @SerializedName("CANCEL") CANCEL,
    @SerializedName("REVIEW") REVIEW,
    @SerializedName("DOWNGRADE") DOWNGRADE,
    @SerializedName("CONSOLIDATE") CONSOLIDATE
}

enum class RecoPriority {
    @SerializedName("HIGH") HIGH,
    @SerializedName("MEDIUM") MEDIUM,
    @SerializedName("LOW") LOW
}

data class RecommendationItem(
    val type: RecoType,
    val priority: RecoPriority,
    val merchant: String,
    val currentCost: Double,
    val potentialSavings: Double,
    val reason: String
)
